src/actions.ts

```diff
@@ -0,0 +1,173 @@
+import {
+  LibraryData,
+  LibrariesData
+} from "./interfaces";
+import DataFetcher from "opds-web-client/lib/DataFetcher";
+import { RequestError, RequestRejector } from "opds-web-client/lib/DataFetcher";
+import BaseActionCreator from "opds-web-client/lib/actions";
+
+/** Create redux actions to be dispatched by connected components, mostly
+    to make requests to the server. */
+export default class ActionCreator extends BaseActionCreator {
+
+  static readonly GET_ALL_LIBRARIES = "GET_ALL_LIBRARIES";
+  static readonly GET_ONE_LIBRARY = "GET_ONE_LIBRARY";
+
+  static readonly EDIT_STAGES = "EDIT_STAGES";
+
+
+  csrfToken: string;
+
+  constructor(fetcher?: DataFetcher, csrfToken?: string) {
+    fetcher = fetcher || new DataFetcher();
+    super(fetcher);
+    csrfToken = csrfToken || null;
+    this.csrfToken = csrfToken;
+  }
+
+  fetchLibraries() {
+    let url = "/admin/libraries";
+    return this.fetchJSON<LibrariesData>(ActionCreator.GET_ALL_LIBRARIES, url).bind(this);
+  }
+
+  fetchLibrary(uuid: string) {
+    let url = "/admin/libraries/" + uuid;
+    return this.fetchJSON<LibraryData>(ActionCreator.GET_ONE_LIBRARY, url).bind(this);
+  }
+
+  editStages(data: FormData) {
+    let url = "/admin/libraries/registration";
+    return this.postForm(ActionCreator.EDIT_STAGES, url, data).bind(this);
+  }
+
+
+
+
+
+
+
+
+
+  postForm(
```

Comment:
I'd prefer if these functions were declared before their invocations in the three functions you have above. Javascript does the hoisting already but better to be explicit.

Reply:
As in...1) create an interface for ActionCreator, 2) literally just move the code, or 3) declare them but not define them (i.e. `let postForm;`), which TypeScript finds upsetting?

Reply:
2)